src/id3/v2/frame/header.rs

```diff
@@ -24,12 +26,20 @@ where
 		return Ok(None);
 	}
 
-	let frame_id = &frame_header[..3];
-	let id_str = std::str::from_utf8(frame_id)
-		.map_err(|_| Id3v2Error::new(Id3v2ErrorKind::BadFrameId(frame_id.to_vec())))?;
-	let id = upgrade_v2(id_str).map_or_else(|| Cow::Owned(id_str.to_owned()), Cow::Borrowed);
-
-	let frame_id = FrameId::new_cow(id)?;
+	let frame_id_bytes = &frame_header[..3];
+	let frame_id = match std::str::from_utf8(frame_id_bytes)
+		.map_err(|_| Id3v2Error::new(Id3v2ErrorKind::BadFrameId(frame_id_bytes.to_vec())).into())
+		.map(|id_str| {
+			upgrade_v2(id_str).map_or_else(|| Cow::Owned(id_str.to_owned()), Cow::Borrowed)
+		})
+		.and_then(FrameId::new_cow)
+	{
+		Ok(id) => id,
+		Err(err) => match parse_mode {
+			ParsingMode::Strict => return Err(err),
+			ParsingMode::BestAttempt | ParsingMode::Relaxed => return Ok(None),
```

Comment:
`Ok(None)` is currently used to indicate that we encountered padding or the reader is exhausted. This won't only ignore the bad frame ID, but stop reading the tag entirely.

I think errors should continue to be thrown from `parse_*_header`, and handled once in `Frame::read`. That'd simplify things a bit, since the match on `parse_mode` appears 3 times in this file.

From there, you can return `Ok((None, false))` from `Frame::read` to indicate that there *are* more frames to come after the failure:
https://github.com/Serial-ATA/lofty-rs/blob/f1e8ab6c2fb1bb275a15d9e4246ad49f2dbbb3da/src/id3/v2/read.rs#L50-L56

Reply:
The `Option` is the wrong choice here and very confusing. It needs to be replaced by a dedicated type that reveals the semantics.